File: 589.n-叉树的前序遍历.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=589 lang=java
 *
 * [589] N 叉树的前序遍历
 */

// @lc code=start
/*
// Definition for a Node.

class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return new ArrayList<>();
        }
        res.add(root.val);
        for (int i = 0; i < root.children.size(); i++) {
            res.addAll(preorder(root.children.get(i)));
        }
        return res;
    }
}
// @lc code=end

File: 589.n-叉树的前序遍历.go
package aaa

/*
 * @lc app=leetcode.cn id=589 lang=golang
 *
 * [589] N 叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Children []*Node
 * }
 */
// type Node struct {
// 	Val      int
// 	Children []*Node
// }
//写法1
// func preorder(root *Node) []int {
// 	res := make([]int, 0)
// 	// 递归调用append方法，需要传入切片的指针，否则返回以前的内存地址
// 	run(root, &res)
// 	return res
// }
// func run(root *Node, res *[]int) {
// 	if root == nil {
// 		return
// 	}
// 	*res = append(*res, root.Val)
// 	if root.Children != nil {
// 		for _, ch := range root.Children {
// 			run(ch, res)
// 		}
// 	}
// }
//写法2 闭包
// func preorder(root *Node) []int {
// 	var res []int
// 	var des func(*Node)
// 	des = func(node *Node) {
// 		if node == nil {
// 			return
// 		}
// 		res = append(res, node.Val)
// 		for _, ch := range node.Children {
// 			des(ch)
// 		}
// 	}
// 	des(root)
// 	return res
// }
//写法3 可变参数
func preorder(root *Node) []int {
	var res []int
	if root == nil {
		return nil
	}
	res = append(res, root.Val)
	for i := 0; i < len(root.Children); i++ {
		res = append(res, preorder(root.Children[i])...)
	}
	return res
}

// @lc code=end
